Various Fixes and enhancements in x86 intrinsics

sayantn opened the pull request:
- Updated the x86-intel.xml to be a more recent update of Intel Intrinsics Guide (v3.6.8)
 - Added functionality to auto-generate a `missing-x86.md` for ease of implementation
 - Updated disassembly to allow windows-gnu targets (they have the same implementation as linux targets, as that uses `objdump` from binutils, and binutils is a dependency of GCC). Add the `x86_64-pc-windows-gnu` target in CI
 - Fixed some of the stream intrinsics.
 - modified floating-point reduce-add and reduce-mul intrinsics to NOT use `simd_reduce_add_unordered` and `simd_reduce_mul_unordered` as Intel specifies a strict associativity. Follow GCC and hand-implement the associativity ourselves (#1533)
 - Fixed `_load_mask32` etc in AVX512BW (they should have taken a `__mmask32`/`__mmask64` pointer, but took `u32`/`u64` pointer)
 - As moves never modify any flags, add `preserves_flags` to the `asm!` blocks for moves
 - Fix `_mm_loadu_si64` (it had target-feature sse, but needs sse2), `_mm256_extract_epi64`, `_mm256_extract_epi32`, `_mm256_cvtsi256_si32` (these had target-feature avx2, but need avx).
 -  Fixed `_mm_cvtt` intrinsics (they were actually calling vcvtss2si, when they should call vcvttss2si)
 - Removed all MMX support from stdarch-verify, and made the target-feature verification stricter
 - Implemented the missing intrinsics mentioned in #1178 with feature-gate `simd-x86-updates` (rust-lang/rust#126936)
 - Fixed `_mm512_kunpackb`
 - Modified the reduce-max and reduce-min intrinsics to preserve associativity specified by Intel and to use the comparison function they described (which is NOT `maxnum` from LLVM)
 - Bumped the OS in CI Docker containers to Ubunto 24.04 (except for in `armv7-unknown-linux-gnueabihf` and `x86_64-unknown-linux-gnu-emulated`)


Modifying fma has been moved to #1597 
Masked load/stores are on standby due to rust-lang/rust#126919

sayantn committed:
Upgraded disassembly to include `windows-gnu` targets

sayantn committed:
Update Intrinsics list

Updated the intrinsics list from version 3.4 to 3.6.8. Added a missing-x86.md file to track progress.

workingjubilee commented on crates/core_arch/src/x86/sse41.rs:
```diff
@@ -1148,6 +1142,35 @@ pub unsafe fn _mm_test_mix_ones_zeros(a: __m128i, mask: __m128i) -> i32 {
+#[inline]
+#[target_feature(enable = "sse4.1")]
+#[cfg_attr(test, assert_instr(movntdqa))]
+#[stable(feature = "simd_x86", since = "1.27.0")]
+pub unsafe fn _mm_stream_load_si128(mem_addr: *const __m128i) -> __m128i {
```
Could you split out the addition of new stable intrinsics under an existing feature flag into a new PR, so that they can be FCPed if necessary?

sayantn replied:
There are 5-6 missed functions in sse2, sse3, ssse3 and avx2. What feature flag should I use?

workingjubilee replied:
Oh, I have no objection to you proposing they be insta-stabilized under that feature flag, it's just better to have insta-stable additions be in their own PR.

sayantn replied:
Oh, ok. So they should be insta-stabilized under `simd_x86`? and what should be the since argument be? (I don't know enough about stabilization)

workingjubilee replied:
The `since` argument should be either the same version as the rest, or the current nightly, I think? I'm not entirely sure either. This is an edge case of a "should've been in the original intrinsic set" thing.

Yes, T-libs-api will look at it and discuss whether or not to simply merge it or whether to FCP it as a formality or what. In general, adding or stabilizing new public API should be split from changing implementation of existing public API, unless someone says something like "let's change this one thing and then stabilize it".

sayantn replied:
Ok, I will send a PR for these missing intrinsics. We cannot use simd_x86 feature with any version other than 1.27.0 (it doesn't compile)

sayantn replied:
Also can we do anything about `_mm_malloc` and `_mm_free`?

workingjubilee replied:
I don't think we should try that here.

sayantn committed:
Fixed many intrinsics

fixed reduce-add and reduce-mul. and load/store of mask32 and mask64. added preserves-flags to mov asm. fixed the missing list. fixed `_mm_loadu_si64`. Added `assert_instr`

sayantn committed:
Fixed some more intrinsics

Added some tests, Fixed incorrect target-features, and verification code for target-features. Removed all MMX support from verification.

sayantn committed:
Add the missing BMI1, SSE2, SSE4.1 and AVX2 intrinsics

sayantn committed:
Update CI to accommodate for windows-gnu targets

sayantn committed:
Fixed `_mm512_kunpackb`, reduce-max and reduce-min

`_mm512_kunpackb` was implemented wrong, and `simd_reduce_max` uses `maxnum` for comparison, which adheres to IEEE754, but Intel specifically says that they do NOT adhere to IEEE754 for NaNs, which can give wrong results

Amanieu commented on crates/core_arch/src/x86/avx2.rs:
```diff
@@ -3124,6 +3137,35 @@ pub unsafe fn _mm256_srlv_epi64(a: __m256i, count: __m256i) -> __m256i {
+///
+/// [Intel's documentation](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm256_stream_load_si256)
+///
+/// # Safety of non-temporal stores
```
This text only applies to non-temporal stores, but this is a non-temporal load which doesn't have weak memory ordering semantics. This whole section can be removed.

Amanieu commented on crates/core_arch/src/x86/sse41.rs:
```diff
@@ -1142,6 +1142,35 @@ pub unsafe fn _mm_test_mix_ones_zeros(a: __m128i, mask: __m128i) -> i32 {
     _mm_testnzc_si128(a, mask)
 }
 
+/// Load 128-bits of integer data from memory into dstt. mem_addr must be aligned on a 16-byte
```
dstt -> dst

Amanieu commented on crates/core_arch/src/x86/sse41.rs:
```diff
@@ -1142,6 +1142,35 @@ pub unsafe fn _mm_test_mix_ones_zeros(a: __m128i, mask: __m128i) -> i32 {
+///
+/// [Intel's documentation](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm_stream_load_si128)
+///
+/// # Safety of non-temporal stores
```
Again, doesn't apply to load.

Amanieu commented on crates/stdarch-verify/tests/x86-intel.rs:
```diff
@@ -284,6 +273,7 @@ fn verify_all_signatures() {
                 "_mm_cvtsi64x_sd",
                 "_bextr2_u64",
                 "_mm_tzcnt_64",
+                "_mm_broadcastsi128_si256",
```
Can you write a quick test for this to avoid needing to add yet another exception to this list?

sayantn replied:
I think it should be on that list - as that is just a list of all aliases. But ok, I will write a test

sayantn committed:
Some fixes as asked by @Amanieu

Amanieu commented on crates/core_arch/src/x86/sse41.rs:
```diff
@@ -1936,4 +1965,14 @@ mod tests {
+    #[simd_test(enable = "sse4.1")]
+    // Miri cannot support this until it is clear how it fits in the Rust memory model
+    // (non-temporal store)
+    #[cfg_attr(miri, ignore)]
```
This is no longer needed.

Amanieu commented on crates/core_arch/src/x86/avx2.rs:
```diff
@@ -5153,6 +5195,16 @@ mod tests {
+    #[simd_test(enable = "avx2")]
+    // Miri cannot support this until it is clear how it fits in the Rust memory model
+    // (non-temporal store)
+    #[cfg_attr(miri, ignore)]
```
Same here.

sayantn committed:
Fixing CI

Fixed x86_64-apple-darwin freezing.
Bump all docker to Ubuntu-24.04 (except for emulated and armv7)

RalfJung commented on crates/core_arch/src/x86/avx2.rs:
```diff
@@ -3124,7 +3137,25 @@ pub unsafe fn _mm256_srlv_epi64(a: __m256i, count: __m256i) -> __m256i {
+/// must be aligned on a 32-byte boundary or a general-protection exception may be generated. To
+/// minimize caching, the data is flagged as non-temporal (unlikely to be used again soon)
+///
+/// [Intel's documentation](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm256_stream_load_si256)
```
This (and all other AVX2 non-temporal operations) should get the same safety comment that the older non-temporal stores have. See e.g. [here](https://doc.rust-lang.org/nightly/core/arch/x86/fn._mm_stream_si128.html).

Amanieu replied:
I checked and only non-temporal *stores* have special memory orderings on x86. x86 non-temporal loads work just like normal loads.

sayantn replied:
@Amanieu told that that doesn't apply to streaming loads, only streaming stores.

RalfJung replied:
Oh, I didn't realize non-temporal loads even are a thing. More nightmare waiting to happen, I guess...

ruriww commented:
> I checked and only non-temporal stores have special memory orderings on x86. x86 non-temporal loads work just like normal loads.

Is there a reason the non-temporal loads are written with inline assembly then? It generates extra instructions when offsetting. https://godbolt.org/z/os8bG3Efr
